rust/src/http2/detect.rs

```diff
@@ -0,0 +1,440 @@
+    match direction {
+        STREAM_TOSERVER => match &tx.type_data {
+            Some(HTTP2FrameTypeData::HEADERS(hd)) => {
+                if (i as usize) < hd.blocks.len() {
+                    let value = &hd.blocks[i as usize].name;
+                    unsafe {
+                        *buffer = value.as_ptr();
+                        *buffer_len = value.len() as u32;
+                    }
+                    return 1;
+                }
+            }
+            _ => {
+                return 0;
+            }
+        },
+        STREAM_TOCLIENT => match &tx.type_data {
+            Some(HTTP2FrameTypeData::HEADERS(hd)) => {
+                if (i as usize) < hd.blocks.len() {
+                    let value = &hd.blocks[i as usize].name;
+                    unsafe {
+                        *buffer = value.as_ptr();
+                        *buffer_len = value.len() as u32;
+                    }
+                    return 1;
+                }
+            }
+            _ => {
+                return 0;
+            }
+        },
+        _ => {}
+    }
+
+    return 0;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_http2_tx_get_header(
+    tx: *mut std::os::raw::c_void,
+    direction: u8,
+    i: u32,
+    buffer: *mut *const u8,
+    buffer_len: *mut u32,
+) -> u8 {
+    let tx = cast_pointer!(tx, HTTP2Transaction);
+    match direction {
+        STREAM_TOSERVER => match &tx.type_data {
+            Some(HTTP2FrameTypeData::HEADERS(hd)) => {
+                if (i as usize) < hd.blocks.len() {
+                    let mut vec = Vec::new();
+                    for j in 0..hd.blocks[i as usize].name.len() {
+                        vec.push(hd.blocks[i as usize].name[j]);
+                        if hd.blocks[i as usize].name[j] == ':' as u8 {
+                            vec.push(':' as u8);
+                        }
+                    }
+                    vec.push(':' as u8);
+                    for j in 0..hd.blocks[i as usize].value.len() {
+                        vec.push(hd.blocks[i as usize].value[j]);
+                        if hd.blocks[i as usize].value[j] == ':' as u8 {
+                            vec.push(':' as u8);
+                        }
+                    }
+                    let value = &vec;
+                    unsafe {
+                        *buffer = value.as_ptr();
+                        *buffer_len = value.len() as u32;
+                    }
+                    return 1;
+                }
+            }
+            _ => {
+                return 0;
+            }
+        },
+        STREAM_TOCLIENT => match &tx.type_data {
+            Some(HTTP2FrameTypeData::HEADERS(hd)) => {
+                if (i as usize) < hd.blocks.len() {
+                    let mut vec = Vec::new();
+                    for j in 0..hd.blocks[i as usize].name.len() {
+                        vec.push(hd.blocks[i as usize].name[j]);
+                        /* this escaping is needed
```

Comment:
Discussion welcomed here

Reply:
The escaping looks fine.  I found it a little faster, if you can determine a maximum size for the output vector to do something like `let mut out = vec![0; size]` then set the indexes directly. You have to make sure you don't overrun the vector though.  You can see where I do this in the JsonBuilder branch for the escaping done there.  In the JsonBuilder I think the performance difference was worth the extra code, I'm not sure about this case.

2 other things about this block of code.  I think `if let Some(HTTP2FrameTypeData::HEADERS(hd)) = &tx.type_data` would be a little cleaner than the match. And second, I'd like to see the escaping moved out of the extern "C" function to lessen the code in unsafe (whether or not an extern "C" is marked unsafe, I consider it unsafe).

Reply:
Thanks

I am going rather with `Vec::with_capacity` to avoid reallocs and keep `push`